i3bar: add support for nonprimary output

mschuwalow opened the pull request:
Adds support for a `nonprimary` output to i3bar. This is useful when configuring your bar using `primary`, as you can have one config for your main display and another for any additional display.

Fixes #4083

wlhlm commented:
Thanks for taking a stab at this. I've been interested in being able to configure i3bar on non-primary outputs for a while.

Associated issues: #4083, https://github.com/i3/i3/issues/2311#issuecomment-213966121

~~Though, looking at your code, it looks quite limited to me: IIUIC, this will pick the first non-primary output and stop. This makes sense if you have two or less outputs, otherwise it will just pick the first non-primary output from the list and that depends on how the list is assembled (not sure if deterministic). I think it makes more sense if the `nonprimary` option displays the bar on all non-primary outputs not just the first one. If you have more then two outputs and want to display the bar only  on a single non-primary output, I thinks it's better to hardcode the output.~~

mschuwalow commented:
As far as I can tell this works with any number of outputs. I'm currently using this with 3 monitors and everything works as expected https://github.com/mschuwalow/nixos-config/blob/master/profiles/desktop/i3.nix#L455.

My understanding of that part of the code is that one map will be sent for each output, so it's enough for us to check whether we are supposed to handle it. In the case of `nonprimary` that means short-circuiting as soon as we see an output that is not marked as primary.

wlhlm commented:
> As far as I can tell this works with any number of outputs. I'm currently using this with 3 monitors and everything works as expected

OK, I misread the code then. Good to hear! 👍

orestisfl commented on docs/userguide:
```diff
@@ -1486,7 +1486,7 @@ directive multiple times.
 *Syntax*:
 ---------------
-output primary|<output>
+output primary|nonprimary|<output>
```
Please briefly explain what nonprimary means. I.e., every output except the primary

orestisfl commented on i3bar/src/outputs.c:
```diff
@@ -195,8 +195,8 @@ static int outputs_end_map_cb(void *params_) {
             if (strcasecmp(params->outputs_walk->name, config.outputs[c]) == 0 ||
-                (strcasecmp(config.outputs[c], "primary") == 0 &&
-                 params->outputs_walk->primary)) {
+                (strcasecmp(config.outputs[c], "primary") == 0 && params->outputs_walk->primary) ||
+                (strcasecmp(config.outputs[c], "nonprimary") == 0 && !params->outputs_walk->primary)) {
```
I think this will be  a bit more readable with a const:
```c
const bool is_primary = params->outputs_walk->primary
```

```suggestion
                (strcasecmp(config.outputs[c], "primary") == 0 && is_primary ||
                (strcasecmp(config.outputs[c], "nonprimary") == 0 && !is_primary)) {
```

orestisfl reviewed (changes requested):
Thanks! Minor changes.

Also, question: Have you verified that the bar location is updated when you change the primary output? It should in theory work.

mschuwalow commented:
> Thanks! Minor changes.
> 
> Also, question: Have you verified that the bar location is updated when you change the primary output? It should in theory work.

Yes, I can confirm that it works. :+1:

orestisfl reviewed (approved):
Great!

Last chore:
1. Add a line in https://github.com/i3/i3/blob/next/RELEASE-NOTES-next that describes the change
2. Squash all commits into one

mschuwalow committed:
i3bar: add support for nonprimary output

Added a new output option `nonprimary` that causes the bar to be
displayed on all outputs except the primary one.

Fixes #4083

mschuwalow commented:
> Great!
> 
> Last chore:
> 
> 1. Add a line in https://github.com/i3/i3/blob/next/RELEASE-NOTES-next that describes the change
> 2. Squash all commits into one

Done. Thanks for guiding me through the process :smile:

orestisfl commented on RELEASE-NOTES-next:
```diff
@@ -28,6 +28,7 @@ working. Please reach out to us in that case!
   • mention rofi in default config file
   • i3-input: add different exit codes for when i3-input fails
   • allow ppt values in move direction and move position commands
+  • i3bar: add “nonprimary” output option
```
Changed the release notes a bit, this is more consistent with how we write them.

orestisfl reviewed (commented):
Thanks!
